src/model_lifecycle.cc

```diff
@@ -0,0 +1,761 @@
+    load_tracker->load_failed_ = true;
+    load_tracker->reason_ +=
+        ("version " + std::to_string(version) + " is at " +
+         ModelReadyStateString(model_info->state_) +
+         " state: " + model_info->state_reason_ + ";");
+  }
+  // Check if all versions are completed and finish the load
+  if (load_tracker->completed_version_cnt_ ==
+      load_tracker->affected_version_cnt_) {
+    // hold 'map_mtx_' as there will be change onto the model info map
+    std::lock_guard<std::recursive_mutex> map_lock(map_mtx_);
+    auto it = map_.find(model_name);
+    // Check if the load is the latest frontground action on the model
+    for (const auto& version_info : it->second) {
+      if (version_info.second->last_update_ns_ >
+          load_tracker->last_update_ns_) {
+        load_tracker->load_failed_ = true;
+        load_tracker->reason_ =
+            "Newer operation has been applied to the model lifecycle, current "
+            "load operation is out-dated.";
+        break;
+      }
+    }
+
+    if (load_tracker->load_failed_) {
+      // Move agent list out of ModelInfo as it needs to be invoked
+      // after all ModelInfos are reset
+      std::shared_ptr<TritonRepoAgentModelList> lagent_list;
+      if (model_info->agent_model_list_) {
+        lagent_list = std::move(model_info->agent_model_list_);
+      }
+      // If any of the versions fails to load, abort the load and unload
+      // all newly loaded versions
+      for (auto& loaded : load_tracker->load_set_) {
+        // Unload directly, the object is being managed either in frontground
+        // or background
+        std::lock_guard<std::recursive_mutex> lock(loaded.second->mtx_);
+        if (loaded.second->model_ != nullptr) {
+          loaded.second->state_ = ModelReadyState::UNLOADING;
+          loaded.second->state_reason_.clear();
+          loaded.second->agent_model_list_.reset();
+          loaded.second->model_.reset();
+        }
+      }
+
+      if (lagent_list) {
+        auto status =
+            lagent_list->InvokeAgentModels(TRITONREPOAGENT_ACTION_LOAD_FAIL);
+        if (!status.IsOk()) {
+          LOG_ERROR << "Agent model returns error on "
+                       "TRITONREPOAGENT_ACTION_LOAD_FAIL: "
+                    << status.AsString();
+        }
+      }
+    } else {
+      // Unload any previous loaded versions that are still available
+      for (auto& version_info : it->second) {
+        auto& mi = version_info.second;
+        std::lock_guard<std::recursive_mutex> info_lk(mi->mtx_);
+        if ((mi->state_ == ModelReadyState::READY) &&
+            (mi->last_update_ns_ < load_tracker->last_update_ns_)) {
+          if ((mi->agent_model_list_ != nullptr) &&
+              (mi->agent_model_list_->LastActionType() ==
+               TRITONREPOAGENT_ACTION_LOAD_COMPLETE)) {
+            auto status = mi->agent_model_list_->InvokeAgentModels(
+                TRITONREPOAGENT_ACTION_UNLOAD);
```

Comment:
Can you explain this part? 

1. General explanation
2. What is the meaning of LastActionType == TRITONREPOAGENT_ACTION_LOAD_COMPLETE
3. Does this unload all previously loaded versions of a model? This sounds like if a new load fails, it will unload the already loaded/running model which is counterintuitive to the fallback logic we discussed for failed loads?

Reply:
Can we make the code something like:
```
// Only need to invoke UNLOAD action if ...
// Otherwise ...
const bool is_first_unload = 
  (mi->agent_model_list_ != nullptr) && 
  (mi->agent_model_list_->LastActionType() == TRITONREPOAGENT_ACTION_LOAD_COMPLETE)

if (is_first_unload) {
  auto status = mi->agent_model_list_->InvokeAgentModels(TRITONREPOAGENT_ACTION_UNLOAD);
}
```

I find this code hard to read/understand at a glance without being very familiar with the implementation